src/analyses/ai.cpp

```diff
@@ -316,11 +316,13 @@ bool ai_baset::visit_edge(
   trace_ptrt p,
   const irep_idt &to_function_id,
   locationt to_l,
+  trace_ptrt caller_history,
```

Comment:
I do have a pretty strong preference for not using `nullptr` to indicate optional arguments – this isn't obvious from the signature, and frequently makes the logic inside more complicated if you need to disambiguate between pointers you're allowed to dereference and ones you aren't...

Also, are these `trace_ptrts` all `shared_ptr`? In that case it'd seem a bit wasteful to me,  unless these are stored somewhere (I can't quite see it, maybe somewhere further down in the call stack)?

Reply:
I have renamed as you suggested earlier.  I hope the invariants in `ai_history.h` show when the optional argument is present.  `step` kind of has to be a case switch because of the different types of instruction.

Yes, they are all `shared_ptr`s.  They are stored in something that implements the `ai_storaget` interface, allowing you to store states and histories with varying degrees of sensitivity (per history, per location, soon per function and just once).  It's not great but I had a solution with `unique_ptr` but it involved way too many copies.